src/client.js

```diff
@@ -67,19 +68,36 @@ exports.buildConnectionString = buildConnectionString;
  */
 class Client extends events.EventEmitter {
   constructor({username, password, hostname, vhost, connectionString, recycleInterval,
-    retirementDelay, minReconnectionInterval, monitor}) {
+    retirementDelay, minReconnectionInterval, monitor, tcCredentials, rootUrl, contact}) {
     super();
 
     if (connectionString) {
       assert(!username, 'Can\'t use `username` along with `connectionString`');
       assert(!password, 'Can\'t use `password` along with `connectionString`');
       assert(!hostname, 'Can\'t use `hostname` along with `connectionString`');
       assert(!vhost, 'Can\'t use `vhost` along with `connectionString`');
+      assert(!tcCredentials, 'Can\'t use `tcCredentials` along with `connectionString`');
+      assert(!rootUrl, 'Can\'t use `rootUrl` along with `connectionString`');
       this.connectionString = connectionString;
 
       // extract the username as namespace
       const connURL = new URL(connectionString);
       this.namespace = decodeURI(connURL.username);
+    } else if (tcCredentials && rootUrl) {
+      assert(username, 'options.username is required');
+      assert(!password, 'Can\'t use `password` along with `tcCredentials`');
+      assert(!hostname, 'Can\'t use `hostname` along with `tcCredentials`');
+      assert(!vhost, 'Can\'t use `vhost` along with `tcCredentials`');
+      this.pulse = taskcluster.Pulse({
+        tcCredentials,
+        rootUrl,
+      });
+      this.namespace = username;
+      this.contact = contact;
+      const pulsecredentials = pulseCredentials();
```

Comment:
This is an async function, so you need to call it with `await`.  But it's too early to call it here -- rather, it should be called just before connecting, and only if necessary.